### PR TITLE
Try: Aggressive TinyMCE deprecation

### PR DESCRIPTION
**Pull request by tyxla**
## What?
This PR attempts to experiment with an aggressive strategy for deprecating TinyMCE. 

I'd love to get some feedback from everyone!

## Why?
The end goal is to fully deprecate the Classic block and all TinyMCE assets because they end up adding a lot of burden to the loading experience. With that approach, eventually, we could fully remove TinyMCE one day.

## How?
The PR suggests that we wrap the functionality in a Gutenberg Experiment so that everyone can enable it to opt-in.

The experiment disables loading any TinyMCE assets and the classic block by default.

Then, if TinyMCE usage is detected, it refreshes the page and adds an additional query argument. On the subsequent page load, all assets are loaded, classic block as well.

That way, a new installation, or an installation that doesn't utilize tinyMCE will never have to load its assets or see the Classic block anymore.

Furthermore, we detect if the post has a classic block in the content and make sure to load it in that case.

Finally, we detect that the user just inputs some raw HTML in their editor, in which case we alter the messaging to suggest the user either convert to an HTML block (from where they can convert to blocks) or refresh the page if they insist on using the Classic block.

## Testing Instructions
* Checkout the PR and build Gutenberg.
* Go to Experiments http://localhost:8888/wp-admin/admin.php?page=gutenberg-experiments.
* Enable the "Disable TinyMCE and Classic block" experiment and click "Save Changes".
* Open the post editor with a new post: http://localhost:8888/wp-admin/post-new.php
* Verify that you don't see the Classic editor in the inserter.
* Verify that you don't see any assets when you search for "tinymce" in the network tab of your browser dev tools (except for the tinymce-proxy.js file, which is expected as part of this experiment)
* Verify that there are no errors in the console. If you are seeing any errors in the console, verify you also see them in `trunk`.
* Head to `/lib/experimental/disable-tinymce.php` and uncomment the 8th line by removing the leading `//`.
* Open the post editor with a new post: http://localhost:8888/wp-admin/post-new.php
* Verify that the page gets refreshed shortly after loading.
* Verify that you now have the Classic block available and it works well when inserted. 
* Verify that after the new page is loaded, you now have the `requiresTinymce` query argument set.
* Disable the experiment from the Experiments page.
* Open the post editor again.
* Verify that Classic block is available and it works well. 
* Delete the query argument manually from your browser's URL address.
* Verify that Classic block is still available and it works well. 
* Start a new post, and use the code editor to write a sentence.
* Change back to the visual editor, and verify you can see a message like on the screenshot below.
* Save the post, refresh the page, and verify you can see the paragraph loaded within the Classic block.

### Testing Instructions for Keyboard
None

## Screenshots or screencast 

The message one sees when they edit their post in "code editor", they add some raw HTML and they switch back to the visual editor:

Screenshot 2023-05-25 at 17 52 41

### CODE REVIEW

File: lib/client-assets.php
Comment (tyxla) [inline lib/client-assets.php]: We're now loading `editor` as a dependency of `wp-block-library` only if the experiment is disabled and the "require tinymce" ~cookie~ query argument isn't present.

File: lib/experimental/assets/tinymce-proxy.js
Comment (tyxla) [inline lib/experimental/assets/tinymce-proxy.js]: Will this be enough? :trollface:

File: lib/experimental/assets/tinymce-proxy.js
Comment (tyxla) [inline lib/experimental/assets/tinymce-proxy.js]: @mtias we talked about displaying a popup to the user, but I reconsidered this as being too aggressive and invasive. With this approach, we're simply detecting usage of TinyMCE, and if there is any, we'll save a cookie and redirect, and on the next page load, the user will have the classic block and TinyMCE.

File: lib/experimental/class--wp-editors.php
Comment (tyxla) [inline lib/experimental/class--wp-editors.php:17]: Much of the TinyMCE assets are hardcoded to be loaded and not enqueued in a pluggable way. So we're creating a new class and requiring it conditionally, in which case it just doesn't load them anymore.

File: lib/experimental/disable-tinymce.php
Comment (tyxla) [inline lib/experimental/disable-tinymce.php]: Right now, this functionality is powered by a Gutenberg experiment. The experiment is supposed to work as a primary on/off button for the tinyMCE deprecation.

File: lib/experimental/disable-tinymce.php
Comment (tyxla) [inline lib/experimental/disable-tinymce.php]: If the cookie is there, that means the user needs it for something, so we're stopping here and continuing to enqueue and load TinyMCE assets.

File: lib/experimental/disable-tinymce.php
Comment (tyxla) [inline lib/experimental/disable-tinymce.php]: Uncomment this line to force usage of TinyMCE. That can be used to test the cookie logic.

File: lib/experimental/editor-settings.php
Comment (tyxla) [inline lib/experimental/editor-settings.php:94]: This is the global variable that we use to pass the experiment to the client side.

File: packages/block-library/src/index.js
Comment (tyxla) [inline packages/block-library/src/index.js]: We're also loading the classic block if:
* The experiment is disabled
* There is a classic block detected in the content of the current post
* The "require tinymce" cookie is set

File: lib/experimental/assets/tinymce-proxy.js
Comment (mtias) [reply lib/experimental/assets/tinymce-proxy.js]: We should be able to detect usage server side on the parse_blocks loop before hitting the client. Isn't that more straightforward?

Comment (tyxla) [reply lib/experimental/assets/tinymce-proxy.js]: Well, I'm not sure about that - last time I checked, `parse_blocks()` would return an empty `blockName` for a classic block instance, and it returns the same if we have an empty space between blocks, too. Also, even if we were able to detect usage of a Classic block that way, we wouldn't be able to detect plugin usage that relies on the `wp.tinymce` global. Thoughts?

Comment (dmsnell) [reply lib/experimental/assets/tinymce-proxy.js]: @tyxla I don't know if that would be a problem on the server for two reasons:

 - there is no "classic block" - it's implicitly "a block without a name" as a way of supporting raw HTML in a post that _isn't_ block material. to that end we can be confident that a block without a name is a classic block because that's exactly the specification we made which determines if something is a classic block.
 - if a block's inner contents is only whitespace then I don't see any reason to consider it a real block, moreso if it's just `\n\n`. the editor ignores these blocks, so we can too on the server. even if a block has thirty newlines, several tabs, and spaces, it's still all one and the same in HTML.

so we could do something like this…

```php
add_filter( 'render_block', function( $html, $block ) {
	if ( empty( $block['blockName'] ) && ! empty( trim( $html ) ) ) {
		wp_enqueue_script( … );
	}

	return $html;
} );
```

Comment (tyxla) [reply lib/experimental/assets/tinymce-proxy.js]: Makes sense to me. I'm happy to try it out for detecting a classic block usage. Stay tuned.

I still think we're still going to need the cookie approach on the client side for instances of plugins that rely on the tinymce globals - there's no way to catch those from the server.

Comment (dmsnell) [reply lib/experimental/assets/tinymce-proxy.js]: what about setting a getter on `window.tinymce` that reloads the page with extra args to request it? no need to opt-in, and for those situations where we guessed wrong while being aggressive, we can correct that with a page refresh

```js
if (!('tinymce' in window)) {
	Object.defineProperty(window, 'tinymce', {
		get() {
			reloadPageWithMCE()
		}
	});
}
```

**Comment by tyxla** [comment]
I haven't managed to make any progress on this one this week, but I'm planning to resume work on it next week.

**Comment by tyxla** [comment]
Update: the PR is in a better state now:

* We're still using an experiment as a flag. I want to keep that since this will likely require some extensive testing with various site builders and custom editor plugins once it's shipped.
* By default, we're still not loading any of TinyMCE's assets and the classic block at all.
* We're still detecting a `wp.tinymce` global usage and refreshing the page to load it with TinyMCE and the classic block.
* We're now properly detecting if the currently edited post has a classic block, and loading TinyMCE with the Classic block accordingly.
* We're now handling the attempts of anyone to add raw HTML through the code editor with a custom message:

Screenshot 2023-05-25 at 17 52 41

After refreshing, the user will see the Classic block as expected. We're using that use case as a last resort to have them convert to an HTML block. At that time they can also convert the HTML to blocks. We could also add a "Convert to blocks" button next to the "Keep as HTML" button if that will help.

I still need to test thoroughly and see if there are any edge cases to fix, but I'm happy to accept any feedback already.

File: lib/experimental/disable-tinymce.php
Comment (tyxla) [inline lib/experimental/disable-tinymce.php:8]: Uncomment this line to force usage of TinyMCE. That can be used to test the ~cookie~ query argument logic.

File: lib/client-assets.php
Comment (tyxla) [inline lib/client-assets.php]: We'll load TinyMCE only if:

* The "Disable TinyMCE" experiment is disabled
* The ~cookie~ query argument is set (the one that's set when 3rd party TinyMCE usage is detected)
* The current post contains an instance of the classic block

**Comment by dmsnell** [comment]
> We're now handling the attempts of anyone to add raw HTML through the code editor with a custom message:

Does this show the HTML preview of the raw HTML or only show the error box? It can be very disorienting when all you get is the error box, because it removes the context of knowing what you're working with.

**Comment by tyxla** [comment]
> Does this show the HTML preview of the raw HTML or only show the error box? It can be very disorienting when all you get is the error box, because it removes the context of knowing what you're working with.

Good question - yes, it shows a preview below the error box - see below.

Typing in the code editor:
Screenshot 2023-05-30 at 16 37 01

Switching to the visual editor:
Screenshot 2023-05-30 at 16 37 07

**Comment by tyxla** [comment]
I'm pretty satisfied with the state of the PR at this point, I think it's a good time to get some more feedback 😉

File: lib/experimental/disable-tinymce.php
Comment (youknowriad) [inline lib/experimental/disable-tinymce.php:19]: Can't we use an "editor" specific hook here like block_assets or thing like that?

File: lib/experimental/editor-settings.php
Comment (youknowriad) [reply lib/experimental/editor-settings.php:94]: What's the thinking here? making this an experiment for some time to see the impact? What about removing the experiment but leaving this as a plugin-only (no backport to core) feature until we're certain.

File: packages/block-library/src/index.js
Comment (youknowriad) [inline packages/block-library/src/index.js]: Can we unify "classic block in current post's content" and "requireTinyMCE" is set? Can we just set the cooking when the current post has classic block for instance?

**Review by youknowriad — COMMENTED** [review]
Is there a way to make the "fallback" block the HTML block instead of the classic block or is this too aggressive for now?

File: lib/experimental/class--wp-editors.php
Comment (youknowriad) [reply lib/experimental/class--wp-editors.php:17]: Not sure I understand this part? How does loading this class prevents TinyMCE from loading?

File: lib/experimental/assets/tinymce-proxy.js
Comment (mcsf) [reply lib/experimental/assets/tinymce-proxy.js:4]: (Ignore my accidental resolution of this thread.)

**Comment by mtias** [comment]
I'm not a fan of the cookie solution. Can we instead look at using the block manager for this? The change would be to disable Classic by default and re-enable (instead of setting a cookie) based on your outlined heuristics:

image

File: lib/experimental/assets/tinymce-proxy.js
Comment (mcsf) [reply lib/experimental/assets/tinymce-proxy.js]: > what about setting a getter on `window.tinymce` that reloads the page with extra args to request it? no need to opt-in, and for those situations where we guessed wrong while being aggressive, we can correct that with a page refresh

I'd like to see this tested too

**Comment by tyxla** [comment]
> I'm not a fan of the cookie solution. Can we instead look at using the block manager for this? The change would be to disable Classic by default and re-enable (instead of setting a cookie) based on your outlined heuristics:

Unfortunately, that won't do the job, since we need to know if we need to load TinyMCE as early as during server-side rendering, and the block manager uses preferences that are saved by default in local storage, and PHP doesn't have access to local storage. 

So I'm going to try the other suggested alternative, a query argument.

File: lib/experimental/assets/tinymce-proxy.js
Comment (tyxla) [reply lib/experimental/assets/tinymce-proxy.js]: Sounds like a plan 👍 I wasn't very fond of the cookie solution either, so while I wanted to try with the block manager, this won't do (see [here](https://github.com/WordPress/gutenberg/pull/50387#issuecomment-1589081558) for more details), so I'll go with a query argument.

The only downside of that is that we might end up needing some extra effort to persist that query argument, but then again, we won't have to worry about cleaning it up either, so it might end up being quite a solid approach.

Will work on it and update y'all.

Comment (tyxla) [reply lib/experimental/assets/tinymce-proxy.js]: Updated in https://github.com/WordPress/gutenberg/pull/50387/commits/896f8cf5a8b993dd409469af249a70fdc97c0467 to use a query argument instead of a cookie. Other than not persisting the value, this approach appears to be working fine - let me know what you think @dmsnell @mcsf

File: lib/client-assets.php
Comment (tyxla) [inline lib/client-assets.php]: We're also loading the classic block and the editor assets if:
* The experiment is disabled
* There is a classic block detected in the content of the current post
* The "require tinymce" query argument is set

File: lib/load.php
Comment (tyxla) [inline lib/load.php:114]: Right now, this functionality is powered by a Gutenberg experiment. The experiment is supposed to work as a primary on/off button for the tinyMCE deprecation.

File: lib/experimental/assets/tinymce-proxy.js
Comment (tyxla) [inline lib/experimental/assets/tinymce-proxy.js:3]: We're using a query argument in order to be able to detect we need TinyMCE from the server.

File: lib/experimental/class--wp-editors.php
Comment (tyxla) [reply lib/experimental/class--wp-editors.php:17]: This, by default, is a core WP class that takes care of generating and "enqueueing" the editor assets. The problem is that for some of the scripts it doesn't use the traditional way to enqueue scripts, rather it will just output a huge blob of them inline, see:

https://github.com/WordPress/wordpress-develop/blob/559e6cecf49505ecc3a21fbff647fdc5c3f892a3/src/wp-includes/class-wp-editor.php#L1567

So the only way to not spit that blob of scripts out is to override this class. This is actually an optimization because none of what it does will even get enqueued that way (look around the class I linked above, it also enqueued some assets), so we won't need to dequeue them either. 

Luckily, redeclaring the class is possible because of the way that we check if the class is declared before being required, for example:

https://github.com/WordPress/wordpress-develop/blob/559e6cecf49505ecc3a21fbff647fdc5c3f892a3/src/wp-admin/includes/ajax-actions.php#L1923

and:

https://github.com/WordPress/wordpress-develop/blob/559e6cecf49505ecc3a21fbff647fdc5c3f892a3/src/wp-admin/includes/deprecated.php#L771

I realize this is a bold approach, but it makes it so straightforward to remove all editor assets that would otherwise be either enqueued or plainly spit out as inline scripts in the admin. This is one of the main reasons why I prefer having this whole thing wrapped as an experiment before releasing it to the public.

File: lib/experimental/editor-settings.php
Comment (tyxla) [reply lib/experimental/editor-settings.php:94]: Yes, I prefer adding it as an experiment just in case, so we can further polish it without breaking the experience for anyone. I've also explained some of the rationale [here](https://github.com/WordPress/gutenberg/pull/50387#discussion_r1227972394). After I have more confidence, I'm happy to remove the experiment wrapper altogether.

File: lib/experimental/disable-tinymce.php
Comment (tyxla) [reply lib/experimental/disable-tinymce.php:19]: No, because we need an action that actually is executed later and prepared to render an inline script. If we render during `enqueue_block_assets`, it's not guaranteed that `window.wp` will be declared.

File: packages/block-library/src/index.js
Comment (tyxla) [reply packages/block-library/src/index.js]: We're no longer working with a cookie, we're rather working with a query argument since https://github.com/WordPress/gutenberg/pull/50387/commits/896f8cf5a8b993dd409469af249a70fdc97c0467, so I think this should no longer be relevant - let me know what you think.

**Comment by tyxla** [comment]
Thanks for the great feedback @youknowriad @dmsnell @mcsf @mcsf! 🙇‍♂️ 

I think all feedback has been addressed and answered. ✅ 

> Is there a way to make the "fallback" block the HTML block instead of the classic block or is this too aggressive for now?

I agree this is a bit too aggressive, but I'd be willing to try it out separately once this lands.

Furthermore, we're no longer using a cookie, but a query argument instead. The rest of the functionality should be working as before - I've updated the documentation and testing instructions accordingly. 📝 

The PR can use another look when you get a chance, thanks! 🙌

**Comment by tyxla** [comment]
This could still use another round of feedback, if anyone has a chance 👀

File: lib/experimental/assets/tinymce-proxy.js
Comment (dmsnell) [inline lib/experimental/assets/tinymce-proxy.js:5]: from what it's doing, this code seems more like `reloadWithTinymce` than `markAsRequiringTinymce` because it actually performs the reload.

File: lib/client-assets.php
Comment (dmsnell) [reply lib/client-assets.php]: would it be easier to fit together if instead of saying something like `requiredTinymce` we passed to PHP something more WP-oriented, like `?wp-dependencies-enque=editor`? just an idea

File: lib/experimental/disable-tinymce.php
Comment (dmsnell) [inline lib/experimental/disable-tinymce.php]: looks like this line was supposed to be deleted at some point but wasn't.

File: lib/experimental/disable-tinymce.php
Comment (dmsnell) [inline lib/experimental/disable-tinymce.php]: another comment on naming. what it's checking isn't particularly what it's saying it is. maybe something like `gutenberg_post_being_edited_requires_classic_editor()`?

File: packages/block-library/src/index.js
Comment (dmsnell) [inline packages/block-library/src/index.js]: might consider rephrasing this; I found tracking some of the negatives difficult, and "either" normally implies one of two

```js
// When in a WordPress context, conditionally
// add the classic block and TinyMCE editor
// under any of the following conditions:
//   - the current post contains a classic block
//   - TinyMCE hasn't been intentionally disabled
//   - a query argument specifies that TinyMCE should be loaded
```

on this point I guess I have two follow-up questions: would these conditions be better as named variables so that the comment doesn't need to encode the logic? should we be loading TinyMCE if `__experimentalDisableTinymce` is set?

```js
const runningInWpContext = window?.wp?.oldEditor;

if ( ! runningInWpContext ) {
	return blocks.filter( Boolean );
}

const postContainsClassicBlock = window?.wp?.needsClassicBlock;
const tinymceNormallyAllowed = ! window?.__experimentalDisableTinymce;
const queryRequiresTinymce = !! new URLSearchParams( window?.location?.search ).get( 'requiresTinymce' );

if ( postContainsClassicBlock || tinymceNormallyAllowed || queryRequiresTinymce ) {
	blocks.push( classic );
}

return blocks.filter( Boolean );
```

**Review by dmsnell — COMMENTED** [review]
Left a few notes on naming and communication. Overall seems worth a try.

File: lib/experimental/assets/tinymce-proxy.js
Comment (tyxla) [reply lib/experimental/assets/tinymce-proxy.js:5]: Nice tip, renamed in https://github.com/WordPress/gutenberg/pull/50387/commits/7821c4b4cf9b8ef619ea621421267453d9702127.

File: lib/client-assets.php
Comment (tyxla) [reply lib/client-assets.php]: I think this would imply we want to do this magic with other groups of assets. Since I'm not sure this will be a necessity, I prefer keeping it simple as per the YAGNI principle and leaving this just as an idea.

File: lib/experimental/disable-tinymce.php
Comment (tyxla) [reply lib/experimental/disable-tinymce.php]: Good catch, rephrased to reflect the current state of things.

File: lib/experimental/disable-tinymce.php
Comment (tyxla) [reply lib/experimental/disable-tinymce.php]: Nice suggestion 👍 I've only changed `editor` to `block` to better reflect the intentions here: `gutenberg_post_being_edited_requires_classic_block`. Done in https://github.com/WordPress/gutenberg/pull/50387/commits/d0ecf5de9cdf356e36311ed48a374b5a414111ad

File: packages/block-library/src/index.js
Comment (tyxla) [reply packages/block-library/src/index.js]: I really appreciate you taking the time to improve the comment 🙌 

In https://github.com/WordPress/gutenberg/pull/50387/commits/1c5199025caab2a9e270be6f92de0ffaa6019f79 I've reused your suggestion and just slightly updated it to clarify that the **experiment** needs to be disabled as one of the conditions, not TinyMCE. 

Then on your suggestion to change the comment to using variables, I think this makes it less readable, and we have enough specific information to warrant having a comment. I think it would be more difficult to convey all that information within variable names. I think the new version of the comment is doing a great job with that 👍

**Comment by tyxla** [comment]
Thanks for the great feedback @dmsnell 🙌 

I've addressed it and left comments with clarifications.

The PR is ready for another look 👀

File: packages/block-library/src/index.js
Comment (dmsnell) [reply packages/block-library/src/index.js]: thanks!

I guess I still find this experiment condition confusing, which is why I had inverted the variable to avoid having to follow the negations. this is just a personal response though; not a demand. each time I read it, I do spend what seems like too long trying to understand exactly what it's trying to check

```
TinyMCE experiment hasn't been intentionally disabled
```

so this is an experiment to use TinyMCE?
if it's disabled, that means don't load the TinyMCE editor, right?

```
! window?.__experimentalDisableTinymce
```

Oh no, wait, if the experiment to _disable_ TinyMCE is _not_ activated…

🤔 

What about this? It's closer to the code check anyhow. Maybe it's just the mention of "experiment" without indicating that the experiment is to disable the editor that leads to confusion.

> Experiment to disable TinyMCE isn't active.

**Comment by dmsnell** [comment]
@tyxla thanks for the changes! beyond a follow-up comment I left I have nothing more to add 👍

File: packages/block-library/src/index.js
Comment (tyxla) [reply packages/block-library/src/index.js]: Nice, @dmsnell, agreed that what you suggested is much better phrasing. 

Updated in 7b3ddaca89f4497581258d5fae66fe46a98c7b41.

Thank you 🙌

**Comment by tyxla** [comment]
> @tyxla thanks for the changes! beyond a follow-up comment I left I have nothing more to add 👍

Thank you @dmsnell, I've addressed that suggestion!

I'm good with it landing the PR because the experimental wrapper makes things much safer to test, play with and improve in future iterations.

I'd say this is ready to go, but I'll wait for someone to approve it specifically before I 🚢 .

**Review by dmsnell — APPROVED** [review]
Let's keep an eye on this and see how it goes 👍

File: lib/experimental/disable-tinymce.php
Comment (tyxla) [inline lib/experimental/disable-tinymce.php:8]: I'm going to keep that for now, just in case we'd like to test 3rd party global usage.

**Comment by mtias** [comment]
Looking forward to seeing the performance impact of this one!

**Comment by tyxla** [comment]
> Looking forward to seeing the performance impact of this one!

Me too! In the meantime, since this is an experiment, I'm planning to run perf tests on it in #52245.

**Comment by cbravobernal** [comment]
We already have this warning on console. Is due to disabling tinymce?

Screenshot 2023-07-03 at 15 06 07

**Comment by tyxla** [comment]
@c4rl0sbr4v0 this is likely related - that's exactly why this work is an experiment, so one could be able to disable it and resolve the warning.

In the meantime, it would be really helpful to know more about:
* On which admin page are you seeing the error?
* What steps did you follow to reproduce the error?
* What WP version are you using?
* What plugins do you have active on your site?

Thanks in advance!

**Comment by cbravobernal** [comment]
> On which admin page are you seeing the error?

On page and post edit.

> What steps did you follow to reproduce the error?

Just opening a console inside a post or a page edit.

> What WP version are you using?

The one from wp-env. 6.3-alpha-56087/

> What plugins do you have active on your site?

Just Gutenberg in trunk.

**Comment by tyxla** [comment]
@c4rl0sbr4v0 thanks for providing more information, however I'm unable to reproduce, so I have a few more questions:

* Are you using the latest `trunk`?
* Have you run `npm run dev` on `npm run build` with that latest `trunk`?
* Is the "Disable TinyMCE and Classic block" experiment turned on or off on your installation?
* Can you share some more thorough error backtrace information? A screenshot of the entire error with the path where it's invoked could be of great help.

**Comment by cbravobernal** [comment]
I tried again this morning and now I'm unable to reproduce too.

> Are you using the latest trunk?

I was, I updated this morning.

> Have you run npm run dev on npm run build with that latest trunk?

Yep

> Is the "Disable TinyMCE and Classic block" experiment turned on or off on your installation?

I didn't know there was such option 😅 . Now I enabled/disabled and the error does not appear.

> Can you share some more thorough error backtrace information? A screenshot of the entire error with the path where it's invoked could be of great help.

I will do if a get the error again. Sorry for the trouble!

**Comment by tyxla** [comment]
Thank you for the extra information @c4rl0sbr4v0, I really appreciate it!

I tried deleting the option altogether to get to the state you were in before, but couldn't repro it even that way.

Please don't hesitate to ping me if it occurs again!

**Comment by tyxla** [comment]
Following up to this with #52325 that makes it more interoperable with some plugins (like the Classic Editor)

File: lib/experimental/disable-tinymce.php
Comment (mcsf) [inline lib/experimental/disable-tinymce.php:24]: By the way, this construction — a condition calling a function defined only further below — sent me down a rabbit hole on PHP's looks-like-[hoisting](https://developer.mozilla.org/en-US/docs/Glossary/Hoisting)-but-isnt-hoisting rules for [defining and referencing functions](https://www.php.net/manual/en/functions.user-defined.php#:~:text=Functions%20need%20not%20be%20defined%20before%20they%20are%20referenced%2C%20except%20when%20a%20function%20is%20conditionally%20defined%20as%20shown%20in%20the%20two%20examples%20below.). 😄

**Pull request merged**
